lib/modules.nix

```diff
@@ -1204,6 +1204,48 @@ let
       ];
     };
 
+  /*
+    Imports a nix file much like the module system would, but also pass extra
+    arguments.
+
+    This solves the problem of being unable to pass a value from the lexical
+    scope to a single module. This occurs when you return a module from a rich
+    scope, such as that of a flake.
+
+    While you could do `import <file> { ... }` yourself, the result
+    is not as good, because `import` is unaware and does not return the location.
+    `importApply` does keep this location, so that it can be used in error
+    messages.
+
+    The nix file must contain a function that returns a module. A module may
+    itself be a function. See the example below.
+
+    This function does not add support for deduplication and `disabledModules`,
+    although that could be achieved by wrapping the returned module. The reason
+    for this is that second argument of `importApply` may not be the same for
+    different invocations of `importApply`. Hence, the file name may not be a
+    key.
```

Comment:
In which case I believe this should be changed. To support the use case of passing in variables from a scope, the deduplication will work just fine. Only the anti-pattern of generating modules would not work then (which is a good thing)

Reply:
Yeah had the same thought. The distinguishing key can just be based on the import tree though, kind of what happens already with the anonymose keys, e.g. "key-0-2" would be the 2nd module in the imports list of the 0th `modules` argument.

Reply:
Relevant issue:
- https://github.com/NixOS/nixpkgs/issues/215496